taskchampion/src/storage/sqlite/schema.rs

```diff
@@ -0,0 +1,209 @@
+}
+
+/// Update to DbVersion(0, 1).
+///
+/// This function can upgrade from any schema before (0, 1), including those of TaskChampion
+/// 0.8 and 0.9, as all of its operations are performed only if the schema element does not
+/// already exist.
+fn upgrade_to_0_1(t: &Transaction) -> Result<()> {
+    let create_tables = vec![
+            "CREATE TABLE IF NOT EXISTS operations (id INTEGER PRIMARY KEY AUTOINCREMENT, data STRING);",
+            "CREATE TABLE IF NOT EXISTS sync_meta (key STRING PRIMARY KEY, value STRING);",
+            "CREATE TABLE IF NOT EXISTS tasks (uuid STRING PRIMARY KEY, data STRING);",
+            "CREATE TABLE IF NOT EXISTS working_set (id INTEGER PRIMARY KEY, uuid STRING);",
+        ];
+    for q in create_tables {
+        t.execute(q, []).context("Creating table")?;
+    }
+
+    // -- At this point the DB schema is that of TaskChampion 0.8.0.
+
+    // Check for and add the `operations.uuid` column.
+    if !has_column(t, "operations", "uuid")? {
+        t.execute(
+            r#"ALTER TABLE operations ADD COLUMN uuid GENERATED ALWAYS AS (
+                coalesce(json_extract(data, "$.Update.uuid"),
+                         json_extract(data, "$.Create.uuid"),
+                         json_extract(data, "$.Delete.uuid"))) VIRTUAL"#,
+            [],
+        )
+        .context("Adding operations.uuid")?;
+
+        t.execute("CREATE INDEX operations_by_uuid ON operations (uuid)", [])
+            .context("Creating operations_by_uuid")?;
+    }
+
+    if !has_column(t, "operations", "synced")? {
+        t.execute(
+            "ALTER TABLE operations ADD COLUMN synced bool DEFAULT false",
+            [],
+        )
+        .context("Adding operations.synced")?;
+
+        t.execute(
+            "CREATE INDEX operations_by_synced ON operations (synced)",
+            [],
+        )
+        .context("Creating operations_by_synced")?;
+    }
+
+    // -- At this point the DB schema is that of TaskChampion 0.9.0.
+
+    create_version_table(t)?;
+
+    set_db_version(t, DbVersion(0, 1))?;
+
+    Ok(())
+}
+
+fn create_version_table(t: &Transaction) -> Result<()> {
+    // The `singleton` column constrains this table to have no more than one row.
+    t.execute(
+        r#"CREATE TABLE IF NOT EXISTS version (
+                singleton INTEGER PRIMARY KEY CHECK (singleton = 0),
+                major INTEGER,
+                minor INTEGER)"#,
+        [],
+    )
+    .context("Creating table")?;
+    Ok(())
+}
+
+/// Get the current DB version, from the `version` table. If the table or row does not exist,
+/// that is considered version (0, 0).
+pub(super) fn get_db_version(con: &Connection) -> Result<DbVersion> {
+    let version: Option<(u32, u32)> = match con
+        .query_row("SELECT major, minor FROM version", [], |r| {
+            Ok((r.get("major")?, r.get("minor")?))
+        })
+        .optional()
+    {
+        Ok(v) => v,
+        Err(rusqlite::Error::SqliteFailure(_, _)) => None,
```

Comment:
Shouldn't a SqliteFailure log a warning?

Reply:
Hm, this is intended to catch a missing table, but we could possibly check that and fail with an error if the table does exist.

Reply:
I've done that, but I can't see a good way to put a Connection into an error state as part of tests. Opening a DB at a non-existent path fails in `open`. Somehow corrupting the DB after opening it seems too fragile a thing to do in a test.